File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/audio/HeadphonesAudioType.java
package com.mapbox.services.android.navigation.v5.navigation.metrics.audio;

import android.content.Context;
import android.media.AudioManager;

class HeadphonesAudioType implements AudioTypeResolver {
  private static final String HEADPHONES = "headphones";
  private AudioTypeResolver chain;

  @Override
  public void nextChain(AudioTypeResolver chain) {
    this.chain = chain;
  }

  @Override
  public String obtainAudioType(Context context) {
    AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
    if (audioManager == null) {
      return "unknown";
    }
    return audioManager.isWiredHeadsetOn() ? HEADPHONES : chain.obtainAudioType(context);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/audio/HeadphonesAudioType.kt
package com.mapbox.services.android.navigation.v5.navigation.metrics.audio

import android.content.Context
import android.media.AudioManager

internal class HeadphonesAudioType : AudioTypeResolver {
    private val HEADPHONES = "headphones"
    private var chain: AudioTypeResolver? = null

    override fun nextChain(chain: AudioTypeResolver) {
        this.chain = chain
    }

    override fun obtainAudioType(context: Context): String {
        val audioManager: AudioManager? = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        return audioManager?.let { audio ->
            when (audio.isWiredHeadsetOn) {
                true -> HEADPHONES
                false -> {
                    chain?.obtainAudioType(context) ?: "unknown"
                }
            }
        } ?: "unknown"
    }
}
